core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java

```diff
@@ -322,34 +324,53 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
             channelPool.connect(channel.remoteAddress(), H1C, sessionPromise);
         } else {
             // Fail all pending responses.
-            final Throwable throwable;
-            if (ctx.channel().hasAttr(PENDING_EXCEPTION)) {
-                throwable = ctx.channel().attr(PENDING_EXCEPTION).get();
+            final HttpResponseDecoder responseDecoder = this.responseDecoder;
+            final Throwable pendingException;
+            if (responseDecoder != null && responseDecoder.hasUnfinishedResponses()) {
+                pendingException = getPendingException(ctx);
+                responseDecoder.failUnfinishedResponses(pendingException);
             } else {
-                throwable = ClosedSessionException.get();
+                pendingException = null;
             }
-            failUnfinishedResponses(throwable);
+
             // Cancel the timeout and reject the sessionPromise just in case the connection has been closed
             // even before the session protocol negotiation is done.
             sessionTimeoutFuture.cancel(false);
-            sessionPromise.tryFailure(throwable);
+            if (!sessionPromise.isDone()) {
+                sessionPromise.tryFailure(pendingException != null ? pendingException
+                                                                   : getPendingException(ctx));
+            }
         }
     }
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        setPendingException(ctx, new ClosedSessionException(cause));
         Exceptions.logIfUnexpected(logger, channel, protocol(), cause);
-        if (channel.isActive()) {
+        if (!(cause instanceof IOException)) {
             ctx.close();
+        } else {
+            // Netty will close the connection automatically on an IOException.
         }
     }
 
-    private void failUnfinishedResponses(Throwable e) {
-        final HttpResponseDecoder responseDecoder = this.responseDecoder;
-        if (responseDecoder == null) {
-            return;
+    private static Throwable getPendingException(ChannelHandlerContext ctx) {
+        if (ctx.channel().hasAttr(PENDING_EXCEPTION)) {
+            return ctx.channel().attr(PENDING_EXCEPTION).get();
         }
 
-        responseDecoder.failUnfinishedResponses(e);
+        return ClosedSessionException.get();
+    }
+
+    static void setPendingException(ChannelHandlerContext ctx, Throwable cause) {
+        final Attribute<Throwable> attr = ctx.channel().attr(PENDING_EXCEPTION);
+        final Throwable previousCause = attr.get();
+        if (previousCause == null) {
+            attr.set(cause);
+        } else if (previousCause != cause) {
+            previousCause.addSuppressed(cause);
```

Comment:
Is there any chance that the `previousCause` is not suppressible?

Reply:
It's possible but we need to preserve the type of the original exception, such as `UnprocessedRequestException`.

Reply:
Yes couldn't think of that. :sweat_smile:

Reply:
Now just logging, instead of suppressing.